mozregression/telemetry.py

```diff
@@ -0,0 +1,55 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import sys
+from multiprocessing import Process
+
+from mozlog import get_proxy_logger
+
+
+LOG = get_proxy_logger('telemetry')
+
+
+def _send_telemetry_ping(variant, appname):
+    from pathlib import Path
+    from glean import Glean, Configuration
+    from glean import (load_metrics,
+                       load_pings)
+    from mozregression import __version__
+
+    mozregression_path = Path.home() / '.mozilla' / 'mozregression'
+    Glean.initialize(
+        application_id="mozregression",
+        application_version=__version__,
+        upload_enabled=True,
+        configuration=Configuration(
+            ping_tag="mozregression-test-tag"
+        ),
+        data_dir=mozregression_path / "data"
+    )
```

Comment:
I would separate out initialization from submitting the ping here.  `Glean.initialize` [should always be called on every startup](https://mozilla.github.io/glean/book/user/general-api.html#initializing-the-glean-sdk) -- whether telemetry is enabled or not -- so that it can take care of sending any pending pings, and the deletion ping can be handled if necessary.

Reply:
Hmm, so Glean submission is handled as a *single* process launched from the main one in which mozregression is running with multiprocessing. In this model, there's no other way than to do initialization + sending in the same method. I think there's a few advantages to this approach:

1. There is no chance that submitting the ping will interfere with mozregression's own operations.
2. If used as a library, the calling process can do its own telemetry -- this is especially a concern with the mach wrapper, which calls into mozregression this way and will probably be doing its own telemetry: https://searchfox.org/mozilla-central/rev/202a285024f174c2d2bf2152d9cba90a03723eab/tools/mach_commands.py#398

Passing a "glean enabled" flag to the process is easy, of course. If I do that, will it operate correctly? I realize this goes against the grain of how Glean was envisioned to work but I'd rather prioritize making sure mozregression operates with no issue for its user.

Reply:
I think for the multiprocessing issue -- which basically boils down to not having Glean block the application -- we probably need to handle that at the Glean level, if not doing it "all in one go" like this.  I think it's possible -- we do this on other platforms, just that Python is extra annoying because with the GIL threading isn't likely to work well and we have to use multiprocessing.  I have some ideas for this already, and this probably pushes them up in priority.  This would negate the need for passing a "glean enabled" flag to the process -- we should be able to handle that kind of thing within the Glean SDK.

The case of mozregression being both a library and an application depending on the context is new, but should fundamentally not be a problem as long as *only* the application calls Glean initialize, given the current design.  When using mozregression as a library, it can still set metrics and send pings even if the library in which it embeds it is the one calling `Glean.initialize`.

So I think the action item is -- I'll start looking at adding built-in multiprocessing to Glean.  If that looks feasible, we could refactor this to separate out init and recording/submitting as discussed.  If that turns into a quagmire, this approach is probably a good workaround for now.

Reply:
In the case of mach won't the application id be different/wrong in the case of the call to `Glean.initialize`? I assume mach will be passing in something like `mach` for this parameter.

Reply:
Indeed that's true.  Up until now, we've considered data as "sent by applications", even if it's from a library embedded in an application.  For example, application-services sync data ends up in tables related to Fenix.  I think in this case it's less useful to think of mozregression data sent from within mach as different from data sent from outside of mach.  This probably will require a little design discussion to get right -- I think we could still have a single Glean instance per application / process, but maybe we allow for pings that are related to a library is directly and not an application id.

Thanks for poking holes in all of our design assumptions ;)

Cc @dexterp37

Reply:
Ok, I'm going to leave this aspect as-is for now.

Reply:
> Thanks for poking holes in all of our design assumptions ;)

Indeed! Thanks @wlach ! (for future reference, @mdboom filed a [bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1624695) about this).